hypothesis-python/src/hypothesis/internal/conjecture/choicetree.py

```diff
@@ -0,0 +1,145 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Most of this work is copyright (C) 2013-2019 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import absolute_import, division, print_function
+
+from collections import defaultdict
+
+from hypothesis.internal.compat import hrange
+
+
+class Chooser(object):
+    """A source of nondeterminism for use in shrink passes."""
+
+    def __init__(self, tree, prefix):
+        self.__prefix = prefix
+        self.__tree = tree
+        self.__node_trail = [tree.root]
+        self.__choices = []
+        self.__finished = False
+
+    def choose(self, values, condition=lambda x: True):
+        """Return some element of values satisfying the condition
+        that will not lead to an exhausted branch, or raise DeadBranch
+        if no such element exist".
+        """
+        assert not self.__finished
+        node = self.__node_trail[-1]
+        if node.size is None:
+            node.size = len(values)
+            node.n = len(values)
+
+        assert node.size > 0 or len(values) == 0
+
+        depth = len(self.__choices)
+
+        if depth < len(self.__prefix):
+            i = self.__prefix[depth]
+            if i >= len(values):
+                i = 0
+        else:
+            i = 0
+
+        count = 0
+        while node.size > 0:
+            count += 1
+            assert count <= len(values)
+            if not node.children[i].exhausted:
+                v = values[i]
+                if condition(v):
+                    self.__choices.append(i)
+                    self.__node_trail.append(node.children[i])
+                    return v
+                else:
+                    node.children[i] = DeadNode
+                    node.size -= 1
+            i = (i + 1) % len(values)
+        raise DeadBranch()
+
+    def finish(self):
+        """Record the decisions made in the underlying tree and return
+        a prefix that can be used for the next Chooser to be used."""
+        self.__finished = True
+        assert len(self.__node_trail) == len(self.__choices) + 1
+
+        next_value = list(self.__choices)
```

Comment:
Minor: It might be clearer to change this to `next_prefix` to match `prefix` (or whatever it ends up being called), to clarify that we're building the choice path that the chooser will try next.